src/lib.rs

```diff
@@ -65,65 +65,137 @@ The bootloader can be configured through a `[package.metadata.bootloader]` table
 for all possible configuration options.
 */
 
-#![cfg_attr(not(feature = "builder"), no_std)]
-#![feature(maybe_uninit_slice)]
-#![feature(step_trait)]
-#![deny(unsafe_op_in_unsafe_fn)]
 #![warn(missing_docs)]
 
-pub use crate::boot_info::BootInfo;
-pub use crate::config::Config;
-
-/// Configuration options for the bootloader.
-mod config;
-
-/// Contains the boot information struct sent by the bootloader to the kernel on startup.
-pub mod boot_info;
-
-/// Contains the actual bootloader implementation ("bootloader as a binary").
-///
-/// Useful for reusing part of the bootloader implementation for other crates.
-///
-/// Only available when the `binary` feature is enabled.
-#[cfg(feature = "binary")]
-pub mod binary;
-
-/// Provides a function to turn a bootloader executable into a disk image.
-///
-/// Used by the `builder` binary. Only available when the `builder` feature is enabled.
-#[cfg(feature = "builder")]
-pub mod disk_image;
-
-#[cfg(all(target_arch = "x86", not(feature = "builder")))]
-compile_error!(
-    "This crate currently does not support 32-bit protected mode. \
-         See https://github.com/rust-osdev/bootloader/issues/70 for more information."
-);
-
-#[cfg(all(
-    not(target_arch = "x86_64"),
-    not(target_arch = "x86"),
-    not(feature = "builder")
-))]
-compile_error!("This crate only supports the x86_64 architecture.");
-
-/// Defines the entry point function.
-///
-/// The function must have the signature `fn(&'static mut BootInfo) -> !`.
-///
-/// This macro just creates a function named `_start`, which the linker will use as the entry
-/// point. The advantage of using this macro instead of providing an own `_start` function is
-/// that the macro ensures that the function and argument types are correct.
-#[macro_export]
-macro_rules! entry_point {
-    ($path:path) => {
-        /// Kernel entry point.
-        #[export_name = "_start"]
-        pub extern "C" fn __impl_start(boot_info: &'static mut $crate::boot_info::BootInfo) -> ! {
-            // validate the signature of the program entry point
-            let f: fn(&'static mut $crate::boot_info::BootInfo) -> ! = $path;
-
-            f(boot_info)
+use anyhow::Context;
+use std::{
+    fs::{self, File},
+    io::{self, Seek},
+    path::Path,
+};
+
+pub fn create_uefi_disk_image(
+    kernel_binary: &Path,
+    out_fat_path: &Path,
+    out_gpt_path: &Path,
+) -> anyhow::Result<()> {
```

Comment:
This reminds me a lot of the [`bootimage`](https://github.com/rust-osdev/bootimage) crate used in  version 0.9 of the bootloader. Are there any advantages to providing this in a library instead of a binary?

Reply:
Yeah, I think there are. The problem with `bootimage` was that you needed to install a separate dependency for building the project. The boot system of `bootloader` v0.10 fixed that, but still required you to run a specific binary of the bootloader dependency for building it (which is quite hacky and cumbersome).

With this library-based approach, things should become much simpler and more configurable at the same time. The idea is that you create a top level `os` crate with a `kernel` sub-crate (as a cargo workspace). The `os` crate then has an artifact dependency on the compiled kernel, and a normal dependency on `bootloader`:

```toml
[build-dependencies]
kernel = { path = "kernel", artifact = "bin", target = "x86_64-unknown-none" }
bootloader = "0.11.0-alpha.0"
```

This allows us to create a `build.rs` script that has access to both the compiled kernel and bootloader. All it needs to do is to call the `create_uefi_disk_image` function to combine them:

```rust
fn main() {
    let kernel_path = {
        let path = std::env::var("CARGO_BIN_FILE_KERNEL").expect("kernel binary not found");
        PathBuf::from(path)
    };
    let out_fat_path = kernel_path.with_extension("fat");
    let out_gpt_path = kernel_path.with_extension("gpt");

    bootloader::create_uefi_disk_image(&kernel_path, &out_fat_path, &out_gpt_path)
        .expect("failed to build UEFI disk image");

    println!("cargo:rustc-env=UEFI_FAT_PATH={}", out_fat_path.display());
    println!("cargo:rustc-env=UEFI_GPT_PATH={}", out_gpt_path.display());
}
```

This way, the full disk image is built using just `cargo build`, without requiring any external builder binaries. The `main.rs` of the top-level `os` crate then has access to the path of the bootable disk image via the env variables, so it just needs to e.g. launch QEMU with it. Since you write the QEMU command yourself, you can easily pass whatever arguments you like. Or, you could use a different virtual machine. Or copy it to some specific place, e.g. an USB stick. 

I also plan to make the interface of the `bootloader` crate more fine-grained. The idea is that you could e.g. just create the FAT partition if you like or even just use the bootloader executable and do the whole image creation manually. We should also support to place additional files on the FAT partition, for example compiled userspace applications (which can be another artifact dependency of the `os` crate).

---

You can see an example of the build system in the `tests` directory. The nice thing about using cargo's OUT_DIRs for everything is that concurrent tests are unproblematic since each boot image is placed in a different directory. Tests are also much faster than before because the bootloader does not need to be recompiled for each test anymore.

Reply:
Thanks for the very thorough answer! This all sounds very promising!